### PR TITLE
RSDEV-371: Fieldmark java client supporting request to get notebooks, records, CSV and ZIP file

### PR DESCRIPTION
**Pull request by rs-nicof**

### CODE REVIEW

File: src/main/java/com/researchspace/fieldmark/client/FieldmarkClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                HttpMethod.GET,
+                new HttpEntity<>(createHttpHeaders(accessToken)),
+                byte[].class)
+            .getBody();
+
+    File fileFromBytes = null;
+
+    ///// save the photos to the filesystem
+    try {
+      fileFromBytes = new File("file.zip");
```
Comment (mKowalski256) [inline src/main/java/com/researchspace/fieldmark/client/FieldmarkClientImpl.java:104]: I'm not sure where all the temp files created by this method (`file.zip` and ones from expanded zip) are ending up, I think would be more robust to generate a temp dir at the start of the method (with java.nio `Files.createTempDirectory`), and write all the files into it.

**Review by mKowalski256 — COMMENTED** [review]
Looks good so far!

File: LICENSE
[the remark this answers is not in the document]
Comment (rs-nicof) [reply LICENSE:1]: Ah rught, I modified now.
Originally I copied it from the Zenodo client, I guess we need to change it there as well then

File: src/main/java/com/researchspace/fieldmark/client/FieldmarkClientImpl.java (hunk 1 above)
Comment (rs-nicof) [reply src/main/java/com/researchspace/fieldmark/client/FieldmarkClientImpl.java]: Yes that was meant to be changed and optimized. 
It is now done 👍

File: src/test/java/com/researchspace/fieldmark/client/FieldmarkClientRealConnectionTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import com.researchspace.fieldmark.model.FieldmarkRecordsCsvExport;
+import com.researchspace.fieldmark.model.FieldmarkRecordsJsonExport;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+@Disabled("We leave the test Disabled so we can potentially run it "
```
Comment (rs-nicof) [inline src/test/java/com/researchspace/fieldmark/client/FieldmarkClientRealConnectionTest.java:15]: typo

File: src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractor.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  @Setter @Getter
+  private String fileName;
+
+  public FieldmarkFileExtractor() {
+    super(null, byte[].class);
+  }
+
+  @Override
+  public String extractStringFieldValue() {
+    throw new UnsupportedOperationException(
```
Comment (rs-fraser) [inline src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractor.java:17]: Both this class and `FieldmarkLocationExtractor` don't support this method. Would be better to remove the method from `FieldmarkTypeExtractor` and instead implement it in the 2 classes where it is used.

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkDatetimeExtractorTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    assertEquals(DATE, underTest.extractDateStringValue());
+    assertEquals(TIME, underTest.extractTimeStringValue());
+  }
+
+  @Test
+  void testExtractFieldThrowException() {
+    ExtractorParseException thrown = assertThrows(
+        ExtractorParseException.class,
+        () -> new FieldmarkDatetimeExtractor(DATE_TIME + ":123456"),
+        "FieldmarkDatetimeExtractor didnt trow the exception, but it was needed"
```
Comment (rs-fraser) [inline src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkDatetimeExtractorTest.java:36]: typo

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractorTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Test
+  void testExtractFieldStringValueThrowException() {
+    FieldmarkFileExtractor underTest = new FieldmarkFileExtractor();
+
+    UnsupportedOperationException thrown = assertThrows(
+        UnsupportedOperationException.class,
+        () -> underTest.extractStringFieldValue(),
+        "FieldmarkFileExtractor didnt trow the exception, but it was needed"
```
Comment (rs-fraser) [inline src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractorTest.java:37]: typo

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkLocationExtractorTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  @Test
+  void testExtractFieldStringValueThrowException() {
+    FieldmarkLocation locationValue = new FieldmarkLocation(
+        new FieldmarkLocationGeometry(List.of(-3D, 5D)));
+    FieldmarkLocationExtractor underTest = new FieldmarkLocationExtractor(locationValue);
+
+    UnsupportedOperationException thrown = assertThrows(
+        UnsupportedOperationException.class,
+        () -> underTest.extractStringFieldValue(),
+        "FieldmarkLocationExtractor didnt trow the exception, but it was needed"
```
Comment (rs-fraser) [inline src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkLocationExtractorTest.java:37]: typo

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractorTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    assertEquals(String.valueOf(Integer.MAX_VALUE), underTest.extractStringFieldValue());
+    assertEquals("Integer", underTest.getSimpleTypeName());
+  }
+
+  @Test
+  void testExtractFieldValueThrowException() {
+    UnsupportedOperationException thrown = assertThrows(
+        UnsupportedOperationException.class,
+        () -> new FieldmarkTypeExtractor<>(Integer.MAX_VALUE, null),
+        "FieldmarkTypeExtractor didnt trow the exception, but it was needed"
```
Comment (rs-fraser) [inline src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractorTest.java:27]: typo

File: src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractor.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  @Setter
+  protected T fieldValue;
+
+  @Getter
+  protected Class<T> fieldType;
+
+
+  public FieldmarkTypeExtractor(T fieldValue, Class<T> fieldType) {
+    if (fieldType == null) {
+      throw new UnsupportedOperationException("FieldType cannot be null");
```
Comment (rs-fraser) [inline src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractor.java:23]: I think `UnsupportedOperationException` is specifically for where a class doesn't support an operation of one of its interfaces. A validation exception or the `ExtractorParseException` would make more sense here.

**Review by rs-fraser — COMMENTED** [review]
Code looks good.

Left a couple of comments. 

It would also be good to update `README.md` with the java version and maven command to run tests, and any other dependencies if there are any.

File: src/main/java/com/researchspace/fieldmark/model/FieldmarkRecordsJsonExport.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            );
+      }
+    }
+  }
+
+  public Optional<String> getFieldType(String fieldName) {
+    return getFirstRecord().getFieldType(fieldName);
+  }
+
+  public Boolean hasFiles() {
```
[the remark this answers is not in the document]
Comment (rs-nicof) [reply src/main/java/com/researchspace/fieldmark/model/FieldmarkRecordsJsonExport.java:77]: good point, changed

File: src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractor.java (hunk 3 above)
Comment (rs-nicof) [reply src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractor.java:17]: I actually need this method primarely on the `FieldmarkTypeExtractor` 
The File and the Location does not support it becasue it does not make sense (the location has 2 values - longitude and latitude -  and file is a file and has not string value)

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkDatetimeExtractorTest.java (hunk 4 above)
Comment (rs-nicof) [reply src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkDatetimeExtractorTest.java:36]: fixed for all now

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractorTest.java (hunk 5 above)
Comment (rs-nicof) [reply src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkFileExtractorTest.java:37]: thanks, fixed for all now

File: src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractorTest.java (hunk 7 above)
Comment (rs-nicof) [reply src/test/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractorTest.java:27]: thanks, fixed for all now

File: src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractor.java (hunk 8 above)
Comment (rs-nicof) [reply src/main/java/com/researchspace/fieldmark/model/utils/FieldmarkTypeExtractor.java:23]: Yes that makes sense, I'm changing it

File: src/main/java/com/researchspace/fieldmark/model/utils/factory/FieldmarkTypeExtractorFactory.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+          }
+          return new FieldmarkLocationExtractor(locationValue);
+        case "BOOL":
+          return new FieldmarkTypeExtractor<>((Boolean) fieldValue, Boolean.class);
+        case "DATE":
+          return new FieldmarkDateExtractor((String) fieldValue);
+        case "DATETIME":
+          return new FieldmarkDatetimeExtractor((String) fieldValue);
+        default:
+//        case "ARRAY":
```
[the remark this answers is not in the document]
Comment (rs-fraser) [reply src/main/java/com/researchspace/fieldmark/model/utils/factory/FieldmarkTypeExtractorFactory.java:53]: @rs-nicof  what should be done with this commented code - will these cases be implemented post-MVP?

**Comment by rs-fraser** [comment]
> Code looks good.
> 
> Left a couple of comments.
> 
> It would also be good to update `README.md` with the java version and maven command to run tests, and any other dependencies if there are any.

@rs-nicof could you add to the README please?

File: src/main/java/com/researchspace/fieldmark/model/utils/factory/FieldmarkTypeExtractorFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rs-nicof) [reply src/main/java/com/researchspace/fieldmark/model/utils/factory/FieldmarkTypeExtractorFactory.java]: Right, deleted the commented code

**Pull request merged**